### ISSUE TITLE
Consider adding @Nullable / @Nonnull annotations to generated services

### ISSUE THREAD

**Issue by robfletcher** [grpc/grpc-java#4724]
Please answer these questions before submitting your issue.

### What version of gRPC are you using?

1.13.2

### What did you expect to see?

It would be very helpful for those of us implementing *ImplBase classes in Kotlin if the parameters of generated `rpc` methods were annotated with `@javax.annotation.Nonnull` as it gives tooling a hint that the signature of an `override` should use the non-nullable form of the parameter types.

Minor, but would make implementing services in Kotlin that bit more seamless.

**Comment by carl-mastrangelo** [comment]
Sorry, but I don't think this is possible.  The generated code is owned by the project owners, not the grpc library.  We can't really add dependencies after the fact.

**Comment by robfletcher** [comment]
I'm not sure I'd follow. This would be to add the annotation on the generated method headers in the *ImplBase class, not the implementing class itself.

**Comment by carl-mastrangelo** [comment]
Right, but now that adds a jsr305 dependency on all those generated classes, which may not be present in everyones build system.

**Comment by robfletcher** [comment]
fair enough.

**Comment by carl-mastrangelo** [comment]
@robfletcher On a second look, we already depend on javax.annotation.Generated, so this *may* be possible to do.   

@ejona86 thoughts?

**Comment by carl-mastrangelo** [comment]
And... on third look, this would increase the jar size because the retention policy is RUNTIME, rather than SOURCE.    I don't know if that size penalty is worth paying, but that would be the reason we could not add it.

**Comment by dapengzhang0** [comment]
And **javax.annotation:javax.annotation-api** (including `javax.annotation.Generated`) and **jsr305** (`@Nullable`) are different dependencies.

**Comment by ejona86** [comment]
CC @zpencer

**Comment by ejona86** [comment]
grpc-core/stub is pulling in jsr305 already, so many gradle/maven users would be able to compile such generated code. However, gradle users using the new-style dependency specification (`implementation`, `api`, etc) will probably have failures on upgrading. Although we could recommend they add the dependency...

**Comment by zpencer** [comment]
List of annotations supported by kotlin: https://kotlinlang.org/docs/reference/java-interop.html#nullability-annotations

**Comment by ejona86** [comment]
API review discussion:
* How does this improve the Kotlin experience? Seems to reduce the number of places `?` would be necessary. So return values (returned method, returned StreamObserver) and parameters to interfaces it is helpful. Note that in some cases only proto-specific APIs are we aware of non-nullness (e.g., `StreamObserver.onNext()` _could_ be passed `null`; for that case we may need Java 8 annotations).
* Why is code gen being singled out? How does this not apply to the entire library?
* Is this already handled for Protobuf messages?
* If we go down this path, we need a checker that verifies as part of the build. But is it possible to support users using nullability checkers without having to drink the kool-aid ourselves? Right now we use the annotations as documentation; we've not changing the default to NonNull, so the Nullable annotations don't "do" anything. If we are going to do this "for real" we would need compile-time checks to notice when the annotations are incorrect, because they will frequently be incorrect.
* There was also talk about whether a Kotlin-specific code generator is in order. I've heard murmorings about wanting coroutine support

**Comment by MarkMarine** [comment]
I would like to see these annotations included for kotlin, I'm writing server side kotlin code using grpc and it would make my life a lot better if they were everywhere in the library, generated code and non-generated code.
Bringing along another dep is inconsequential compared to the code elegance that would be gained

**Comment by carl-mastrangelo** [comment]
@MarkMarine Could you show an example snippet?  Pretty much every method and every parameter would need to be annotated `@Nonnull`, which would be excessively verbose.

**Comment by mkobit** [comment]
Would definitely appreciate this for Kotlin (and Java as well) - our assumption has been that the parameters would never be `null` in the first place, so we have been explicitly using nonnull-types in our implementations by changing the platform type that IDEA generates by default when we go to implement the RPC method (from say `SayHelloRequest?` to `SayHelloRequest`).

I think there are a couple aspects of this:

1. Generated code should have null constraints annotated
2. gRPC library code itself should have null constraints annotated

This issue covers the generated code. I can't find an issue for the gRPC library code, but there is at least the 1 MR linked above (https://github.com/grpc/grpc-java/pull/5069) so there is some interest in it be present in the library code, too. For the generated code, maybe an experimental generator option can be feature toggled so it is opt-in rather than opt-out? Just throwing it out there there as I don't know if that fits in the general design of protobuf/gRPC.

An implementation note worth keeping in mind for `@Nullable` and `@Nonnull` is the dormant nature of JSR305. Instead, something like https://checkerframework.org/ might be desired. The `@javax.annotation.Generated` problem exists for Java 11 users so it would be nice not to have a similar issue with those annotations.

**Comment by jontro** [comment]
At least for grpc-api it seems that jsr305 is included from com.google.code.findbugs already.  Would be nice if it could be used there. Maybe this belongs in a separate issue

**Comment by Sineaggi** [comment]
Now that jspecify 1.0.0 has been released, should we reconsider adding non-null annotations by default to generated classes?
